Public provider curation api docs (#1951)
hashicorp/terraform-website: content/source/docs/cloud/api/providers.html.md

[404]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404
[409]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/409
[412]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/412
[422]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422
[429]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429
[500]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/500
[504]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/504
[JSON API document]: /docs/cloud/api/index.html#json-api-documents
[JSON API error object]: https://jsonapi.org/format/#error-objects

# Registry Providers API

## Terraform Cloud Registry Implementation

For publicly curated providers, the Terraform Cloud Registry acts as a proxy to the [Terraform Registry](https://registry.terraform.io) for the following:

- List available versions for a specific provider
- Get a specific provider
- Get the latest version for a specific provider

- The public registry discovery endpoints have the path prefix provided in the [discovery document](../../registry/api.html#service-discovery) which is currently `/api/registry/public/v1`.

> - [Authentication](./index.html#authentication) is handled the same as all other Terraform Cloud endpoints.

Author: I went ahead and deleted for now, we don't publish the public registry provider endpoints.

### Sample Proxy Request (public provider)

List available versions for the `aws` provider on the namespace `hashicorp`:

```shell
$ curl https://registry.terraform.io/v2/providers/hashicorp/aws/versions
```

The same request for the same provider on the Terraform Cloud private registry:

```shell
$ curl \
  --header "Authorization: Bearer $TOKEN" \
  https://app.terraform.io/api/registry/public/v2/providers/hashicorp/aws/versions
```

## List Registry Providers for an Organization

`GET /organizations/:organization_name/registry-providers`

Parameter            | Description
---------------------|------------
`:organization_name` | The name of the organization to list available providers from.

> Lists the providers that are available to a given organization.

Author: thanks - I have updated.

Status  | Response                                           | Reason
--------|----------------------------------------------------|----------
[200][] | [JSON API document][] (`type: "registry-providers"`) | The request was successful
[404][] | [JSON API error object][]                          | Providers not found or user unauthorized to perform action

### Query Parameters

This endpoint supports pagination [with standard URL query parameters](./index.html#query-parameters); remember to percent-encode `[` as `%5B` and `]` as `%5D` if your tooling doesn't automatically encode URLs.

Parameter                           | Description
------------------------------------|---------------------

> `q`                                 | **Optional.** A search query string.  Providers are searchable by name, namespace fields.

Author: thanks - I have updated.

    "first": "https://app.terraform.io/api/v2/organizations/my-organization/registry-providers?page%5Bnumber%5D=1&page%5Bsize%5D=6",
    "prev": null,
    "next": "https://app.terraform.io/api/v2/organizations/my-organization/registry-providers?page%5Bnumber%5D=2&page%5Bsize%5D=6",
    "last": "https://app.terraform.io/api/v2/organizations/my-organization/registry-providers?page%5Bnumber%5D=29&page%5Bsize%5D=6"
  },
  "meta": {
    "pagination": {
      "current-page": 1,
      "page-size": 6,
      "prev-page": null,
      "next-page": 2,
      "total-pages": 29,
      "total-count": 169
    }
  }
}
```

## Create a Provider

 `POST /organizations/:organization_name/registry-providers`

Parameter            | Description
---------------------|------------
`:organization_name` | The name of the organization to create a provider in. The organization must already exist, and the token authenticating the API request must belong to the "owners" team or a member of the "owners" team.

[permissions-citation]: #intentionally-unused---keep-for-maintainers

#### Public providers

> When created, the public provider record will be available in the organization's registry provider list. You cannot create versions for public providers as they are maintained in the public registry.

Reviewer: ```suggestion
When created, the public provider record will be available in the organization's registry provider list. You cannot create versions for public providers; you must use the versions available on the Terraform Registry.
```

```

### Sample Response (public provider)

```json
{
  "data": {
    "id": "prov-fZn7uHu99ZCpAKZJ",
    "type": "registry-providers",
    "attributes": {
      "name": "aws",
      "namespace": "hashicorp",
      "registry-name": "public",
      "created-at": "2020-07-09T19:36:56.288Z",
      "updated-at": "2020-07-09T20:16:20.538Z",
      "permissions": {
        "can-delete": true
      }
    },
    "relationships": {
      "organization": {
        "data": {
          "id": "my-organization",
          "type": "organizations"
        }
      }
    },
    "links": {
      "self": "/api/v2/organizations/my-organization/registry-providers/public/hashicorp/aws"
    }
  }
}
```

## Delete a Provider
* `DELETE /organizations/:organization_name/registry-providers/:registry_name/:namespace/:name`

### Parameters

Parameter            | Description
---------------------|------------
`:organization_name` | The name of the organization to delete a provider from. The organization must already exist, and the token authenticating the API request must belong to the "owners" team or a member of the "owners" team.

> `:namespace`         | The provider namespace that the deletion will affect.
> `:name`              | The provider name that the deletion will affect.

Reviewer: ```suggestion
`:namespace`         | The namespace of the provider that will be deleted.
`:name`              | The name of the provider that will be deleted.
```

`:registry_name`     | `public`

[permissions-citation]: #intentionally-unused---keep-for-maintainers

Status  | Response                                             | Reason
--------|------------------------------------------------------|-------
[204][] | Nothing                                              | Success
[403][] | [JSON API error object][]                            | Forbidden - public provider curation disabled

> [404][] | [JSON API error object][]                            | Provider not found or user not authorized

Reviewer: ```suggestion
[404][] | [JSON API error object][]                            | Provider not found or user not authorized to perform action
```
